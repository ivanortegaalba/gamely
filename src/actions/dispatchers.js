import * as actionCreators from './actionCreators';

export const fetchGames = () => dispatch => {
  dispatch(actionCreators.fetchGames());
};

const alreadyExistAsFavourite = (state, shortName) =>
  state.favourites.indexOf(shortName) !== -1;

export const addToFavourites = shortName => (dispatch, getState) => {
  if (!alreadyExistAsFavourite(getState(), shortName)) {
    dispatch(actionCreators.addToFavouritesUnsafe(shortName));
  }
};

export const removeFromFavourites = shortName => (dispatch, getState) => {
  if (alreadyExistAsFavourite(getState(), shortName)) {
    dispatch(actionCreators.removeFromFavouritesUnsafe(shortName));
  }
};

export const filterGameBy = gameName => dispatch => {
  dispatch(actionCreators.filterGamesBy(gameName));
};
