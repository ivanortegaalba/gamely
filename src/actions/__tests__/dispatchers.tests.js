import gameMock from '../../__mocks__/game'
import { addToFavourites, fetchGames, filterGameBy, removeFromFavourites } from '../dispatchers'

describe('dispatchers', () => {
  describe('fetchGames', () => {
    const dispatch = jest.fn()
    it('should dispatch the api call action to fetch games', () => {
      fetchGames()(dispatch)
      expect(dispatch.mock.calls.length).toBe(1)
      expect(dispatch.mock.calls[0]).toMatchSnapshot()
    })
  })

  describe('addToFavourites', () => {
    const dispatch = jest.fn()
    describe('the game is already in favourites', () => {
      const getState = jest.fn().mockReturnValue({ favourites: [gameMock.short] })
      it('should not dispatch the action to save a game as favourite', () => {
        addToFavourites(gameMock.short)(dispatch, getState)
        expect(dispatch.mock.calls.length).toBe(0)
      })
    })
    describe('the game is not in favourites yet', () => {
      it('should dispatch the action to save a game as favourite', () => {
        const getState = jest.fn().mockReturnValue({ favourites: [] })
        addToFavourites(gameMock.short)(dispatch, getState)
        expect(dispatch.mock.calls.length).toBe(1)
        expect(dispatch.mock.calls[0]).toMatchSnapshot()
      })
    })
  })

  describe('removeFromFavourites', () => {
    const dispatch = jest.fn()
    describe('the game is not in favourites', () => {
      const getState = jest.fn().mockReturnValue({ favourites: [] })
      it('should not dispatch the action to remove a game as favourite', () => {
        removeFromFavourites(gameMock.short)(dispatch, getState)
        expect(dispatch.mock.calls.length).toBe(0)
      })
    })
    describe('the game is in favourites', () => {
      it('should dispatch the action to remove a game as favourite', () => {
        const getState = jest.fn().mockReturnValue({ favourites: [gameMock.short] })
        removeFromFavourites(gameMock.short)(dispatch, getState)
        expect(dispatch.mock.calls.length).toBe(1)
        expect(dispatch.mock.calls[0]).toMatchSnapshot()
      })
    })
  })

  describe('filterGamesBy', () => {
    const dispatch = jest.fn()
    it('should dispatch the action to change the filter value at the store', () => {
      const newValueToFilterBy = gameMock.short[0]
      filterGameBy(newValueToFilterBy)(dispatch)
      expect(dispatch.mock.calls.length).toBe(1)
      expect(dispatch.mock.calls[0]).toMatchSnapshot()
    })
  })
})
