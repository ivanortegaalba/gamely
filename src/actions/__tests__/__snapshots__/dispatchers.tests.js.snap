// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dispatchers addToFavourites the game is not in favourites yet should dispatch the action to save a game as favourite 1`] = `
Array [
  Object {
    "shortName": "eightballpool",
    "type": "ADD_TO_FAVOURITES",
  },
]
`;

exports[`dispatchers fetchGames should dispatch the api call action to fetch games 1`] = `
Array [
  Object {
    "Call API": Object {
      "endpoint": "/resources/games.json",
      "schema": ArraySchema {
        "schema": EntitySchema {
          "_getId": [Function],
          "_idAttribute": "short",
          "_key": "games",
          "_mergeStrategy": [Function],
          "_processStrategy": [Function],
          "schema": Object {},
        },
      },
      "types": Array [
        "GAMES_REQUEST",
        "GAMES_SUCCESS",
        "GAMES_FAILURE",
      ],
    },
  },
]
`;

exports[`dispatchers filterGamesBy should dispatch the action to change the filter value at the store 1`] = `
Array [
  Object {
    "filterBy": "e",
    "type": "FILTER_BY",
  },
]
`;

exports[`dispatchers removeFromFavourites the game is in favourites should dispatch the action to remove a game as favourite 1`] = `
Array [
  Object {
    "shortName": "eightballpool",
    "type": "REMOVE_FROM_FAVOURITES",
  },
]
`;
